Add licenses for 3rd party libraries

hroncok opened the pull request:
Given the vendor.txt file, I could download all sdists and extract
those from them. Only 2 were mising, added from GitHub:

https://github.com/gsnedders/python-webencodings/blob/v0.5.1/LICENSE
https://github.com/avakar/pytoml/blob/v0.1.14/LICENSE

See https://github.com/pypa/pipenv/issues/1961

hroncok commented:
I suppose https://github.com/pypa/pip/blob/master/tasks/vendoring/__init__.py needs updating as well :(

pradyunsg commented:
Yeah. I'd prefer that the vendoring script learns how to extract licenses rather than having to do it manually.

hroncok commented:
Let me look at that later today.

hroncok commented:
I've added a "works for me" code. It's very rough and will need some cleaning, but I've pushed it so others can try it as well and send feedback.

hroncok commented on tasks/vendoring/__init__.py:
```diff
@@ -137,6 +147,127 @@ def vendor(ctx, vendor_dir):
+
+
+def license_fallback(vendor_dir, sdist, tar):
+    """A very hackish way to get LICENSE from github
```
Given the ugliness of this, I think maybe hardcoding the 2 license URLs would work better.

hroncok replied:
Done in a fixup commit.

hroncok commented:
Travis CI fails on Python 2, it reports perfectly valid Python 3 only file to have SyntaxError, how much should I dance around that?

pfmoore commented on tasks/vendoring/__init__.py:
```diff
@@ -137,6 +154,105 @@ def vendor(ctx, vendor_dir):
+    """Hardcoded license URLs. Check when updating if those are still needed"""
+    for libname, url in HARDCODED_LICENSE_URLS.items():
+        if libname in sdist_name:
+            *_, name = url.rpartition('/')
```
`_, _, name = url.rpartition('/')` is Python 2 compatible.

pfmoore replied:
It is a Python 3 only script, but we don't exclude it from the Python 2 linter run, for better or worse. It's no big deal - I actually think the explicit version is clearer, anyway.

hroncok replied:
It's not a big deal, I agree.

hroncok commented:
What can I make to move this forward? Thanks

pradyunsg commented:
Aside, as a personal rule of thumb, I start using classes or split my scripts if they go over 250 lines. I'll let @pypa/pip-committers decide if that's something we want here.

hroncok commented:
Note that currently my `pip download` hangs after lockfile, so I cannot test this properly, however it works without lockfile. I've split it into two commits. I've changed all the little things @pradyunsg requested.

If the LICESNE files destination should differ from my implementation, tell me where to put them and I'll change it.

I don't think classes/files would help readability here, but if they are requested, I can use them.

hroncok commented:
Ok, unhanged. Works. I only see this in diff, that I'm confident is not my fault:

```patch
diff --git a/src/pip/_vendor/urllib3/util/ssl_.py b/src/pip/_vendor/urllib3/util/ssl_.py
index dafc75b5..87059434 100644
--- a/src/pip/_vendor/urllib3/util/ssl_.py
+++ b/src/pip/_vendor/urllib3/util/ssl_.py
@@ -3,7 +3,7 @@ import errno
 import warnings
 import hmac
 
-from binascii import hexlify, unhexlify
+from pip._vendor.binascii import hexlify, unhexlify
 from hashlib import md5, sha1, sha256
 
 from ..exceptions import SSLError, InsecurePlatformWarning, SNIMissingWarning
```

pfmoore commented:
I was getting that binascii line when I ran revendor for the release. I manually fixed it then, as I wasn't able to work out what was going on at the time. Maybe I forgot to raise an issue for it - I don't remember now. But yes, it's not related to your changes.

pradyunsg commented:
> That was my deliberate decision to honor the fact that some libs have their own directory and some don't.

I don't really know much about this stuff and so, I'll defer to others here.

pradyunsg commented on tasks/vendoring/__init__.py:
```diff
@@ -137,6 +155,106 @@ def vendor(ctx, vendor_dir):
+    url = HARDCODED_LICENSE_URLS[libname]
+    _, _, name = url.rpartition('/')
+    dest = license_destination(vendor_dir, libname, name)
+    r = requests.get(url, allow_redirects=True)
```
IIRC, requests.get does the fetching, so, that would mean the download would finish and then the downloading message would show up.

hroncok replied:
I.e. the log line should be before the get line? IS that what this comment is about?

hroncok replied:
dcb2e9c

pradyunsg commented on tasks/vendoring/__init__.py:
```diff
@@ -137,6 +155,106 @@ def vendor(ctx, vendor_dir):
+        if part[0].isdigit():
+            break
+        parts.append(part)
+    return'-'.join(parts)
```
A space after return.

Huh. CI should catch stuff like this! Anyways, that's something I'll fix later.

hroncok replied:
it should, right? but it's still orange.

hroncok replied:
ee66bfd

hroncok commented:
Rebased on tom of #5347 fix. Works perfectly for me.

hroncok commented:
Will autosquash the fixups and rebase on master.

xavfernandez commented on tasks/vendoring/__init__.py:
```diff
@@ -141,6 +159,106 @@ def vendor(ctx, vendor_dir):
+    log('Downloading licenses')
+    tmp_dir = vendor_dir / '__tmp__'
+    ctx.run(
+        'pip download -r {0}/vendor.txt --no-binary :all: -d {1}'.format(
```
You could add `--no-deps`.

hroncok replied:
Does it download deps without that?

xavfernandez replied:
No, but we don't vendor dependencies either (cf https://github.com/pypa/pip/blob/b6bbabe3e7bc633f7af9da35f55d8c9d0d7a3f09/tasks/vendoring/__init__.py#L105)

hroncok replied:
Well I (wrongly) assumed `install` would bring dependencies, but `download` would not. Will add `--no-deps`.

xavfernandez commented on tasks/vendoring/__init__.py:
```diff
@@ -164,7 +164,8 @@ def download_licenses(ctx, vendor_dir):
     ctx.run(
-        'pip download -r {0}/vendor.txt --no-binary :all: -d {1}'.format(
+        'pip download -r {0}/vendor.txt --no-binary '
+        '--no-deps :all: -d {1}'.format(
```
err, I was going to merge this but this ended in the wrong commit (I guess you must have amended this change into the last commit :)
And more importantly, `--no-binary --no-deps :all`  won't work, you need `--no-binary :all --no-deps`

hroncok replied:
changed.

xavfernandez commented:
Thanks @hroncok !
